Implementation of RFC: Async Helpers (#3724)

cjtenny opened the pull request:
This implements the `harness/asyncHelpers.js` file proposed in @ptomato 's [RFC](https://github.com/tc39/test262/pull/3724), with its functions `asyncTest(testFunc)` and `assert.throwsAsync(expectedErrorConstructor, funcOrThenable, message)`. Also included are tests for both, including tests mirroring `assert.throws` and a number of additional throwsAsync-specific tests.

Per @jugglinmike 's comments in the maintainers call, never uses await; always explicitly invokes as a thenable.

Implements the original intention of accepting a function or thenable for throwsAsync, happy to modify if there's some general agreement on changing that.

ljharb commented on harness/asyncHelpers.js:
```diff
@@ -0,0 +1,122 @@
+    return Promise.resolve();
+  }
+  try {
+    return testFunc().then(
+      function () {
+        $DONE();
+      },
+      function (error) {
+        $DONE(error);
+      }
+    );
```
what happens if this is trying to test Promise.prototype.then itself?

cjtenny replied:
Good question. Looking at those tests, many of them wouldn't benefit from this function. But also, Promise.prototype.then seems core enough to async functionality that it seems there's a bit of a circular dependency to using both; perhaps that's one of the places it makes sense to continue to use custom logic around manually invoking $DONE? Very open to other ideas, if you have suggestions for how to improve this case.

Either way, I think it remains useful for simplifying tests of other async features not implementing core async interfaces.

cjtenny replied:
Thinking further about it: This only depends on the test function returning a thenable. There are a few options:
1. Do not use asyncTest when testing Promise.prototype.then. Instead, write your own logic for calling $DONE.
2. In your test for Promise.prototype.then, return a custom thenable rather than a Promise. This seems reasonable for a test of Promises anyway.
3. If testing a polyfilled Promise.prototype.then, one would expect the regular tests for Promise.prototype.then to uncover bugs, or the consistency tests for this harnenss to catch them. This itself is not a test for Promise.prototype.then, and doesn't reference or depend on it in any way, as it only references a thenable.

I'll push a quick update removing the references to Promise.resolve and instead returning a simple thenable, so that now the implementation doesn't reference Promise at all.

Does that cover all the cases you were thinking of, @ljharb ?

ljharb commented on harness/asyncHelpers.js:
```diff
@@ -0,0 +1,122 @@
+    );
+  } catch (syncError) {
+    $DONE(syncError);
+    return Promise.resolve();
```
why not make the containing function `async` rather than needing to always return a Promise?

cjtenny replied:
The goal is to ensure that if it's called without the async flag, it will still execute to throw the error on line 11 - I believe top-level calls to async functions might not execute in most runners before the implementation returns, leading to silent failures if an author omits the 'async' flag. But I could well be wrong!

ljharb replied:
You're correct, but i'd expect that to be handled by the unhandled rejection handler, rather than a confusing sync exception from a promise-returning function (ie, zalgo)

cjtenny replied:
The RFC itself [makes the case](https://github.com/tc39/test262/pull/3724/files#diff-d4e487dcd72bb7cd7462c9709f50abd806977f296306b5f4eb77e3fc011ede51R112) for summoning Zalgo here, and test262 runners either don't handle unhandled rejections or handle them without clear messages. Engine262 and JSC both pass a test with an unhandled rejection; v8 will fail it with no output. Node gives an UnhandledPromiseRejection, as does SpiderMonkey. Because of these reasons, I think it's actually worth the top-level Zalgo - hopefully we only would expect this to be called top level. I'd argue it's at least moderately easier to review and catch than the possible mistakes that arise from not having it.

ljharb commented on harness/asyncHelpers.js:
```diff
@@ -0,0 +1,122 @@
+          "Expected to obtain an inner promise that would reject with a" +
+          expectedErrorConstructor.name +
+          " but result was not a thenable";
+        return Promise.reject(new Test262Error(message));
```
same here; why not throw this inside an async function?

cjtenny replied:
You're right, this one could and should be inside an async function - I see no reason it couldn't, anyway. I'll refactor, thanks! I think my brain was stuck on the asyncTest logic and I forgot to re-evaluate here.

ptomato commented on test/harness/asyncHelpers-asyncTest-func-throws-sync.js:
```diff
@@ -0,0 +1,16 @@
+  assert(error instanceof Test262Error);
+}
+asyncTest(function () {
+  throw new Test262Error("Should not be rethrown");
```
Maybe we should make this a string or something like that, so it's easy to assert that it's the exact same thrown value that is passed to `$DONE`?

ptomato commented on test/harness/asyncHelpers-asyncTest-return-not-thenable.js:
```diff
@@ -0,0 +1,29 @@
+
+const doneValues = [];
+function $DONE(error) {
+  doneValues.push(error instanceof TypeError);
```
The RFC doesn't say what kind of error this is, but I'd guess it should be a `Test262Error` in parallel with `assert.throwsAsync()`.

ptomato replied:
I'm not sure I get where this TypeError is thrown, it seems to me that it should be a Test262Error according to the following code:
```js
innerThenable = func();
if (
  innerThenable === null ||
  typeof innerThenable !== "object" ||
  typeof innerThenable.then !== "function"
) {
  message += /* etc */;
  throw new Test262Error(message);
}
```
it seems to me that e.g. `asyncTest(() => 42)` assigns 42 to `innerThenable`, so the condition should trigger here. But the test passes, so I'm not sure what's going on.

cjtenny replied:
> The RFC doesn't say what kind of error this is, but I'd guess it should be a `Test262Error` in parallel with `assert.throwsAsync()`.

Oops, I forgot to respond to this rather than just mark the issue resolved, sorry! That is very much my mistake. Thank you for bringing it up again. It comes from [this line](https://github.com/tc39/test262/blob/main/rfcs/async-helpers.md?plain=1#L103) - this occurs when calling testFunc synchronously throws. In this case, because testFunc is not a function, trying to call it throws a TypeError. The test passes as a result, and if we want to change it to throw a Test262Error I'd support that - but we should change the RFC text to match.

ptomato replied:
I see where I was confused now; I was looking at the code for `assert.throwsAsync()`, but this is `asyncTest()`.

ptomato commented on test/harness/asyncHelpers-throwsAsync-funcOrThenable-throws-sync.js:
```diff
@@ -0,0 +1,63 @@
+  await checkRejects(function () {
+    throw new Test262Error();
+  });
+  await checkRejects({
+    then: function () {
+      throw new Error();
+    },
+  });
+  await checkRejects({
+    then: function () {
+      throw new Test262Error();
+    },
+  });
```
This needs to be updated since we changed `throwsAsync` to just take a function and not a thenable. Also suggest renaming this test.

ptomato commented on test/harness/asyncHelpers-throwsAsync-invalid-funcOrThenable.js:
```diff
@@ -0,0 +1,71 @@
+  await checkRejects("string");
+  await checkRejects(10);
+  await checkRejects();
+  await checkRejects({ then: null });
+  await checkRejects({ then: {} });
+  await checkRejects({ then: "string" });
+  await checkRejects({ then: 10 });
+  await checkRejects({ then: undefined });
```
These are still correct, but can probably be removed as they're no longer relevant if we don't accept a thenable. Also suggest renaming this test.

cjtenny replied:
I realized this test also no longer correctly tests things because it doesn't properly check calls to $DONE - oops. I'll keep one example of a thenable, just to assert that we no longer take thenables.

ptomato commented on test/harness/asyncHelpers-throwsAsync-same-realm.js:
```diff
@@ -0,0 +1,37 @@
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+description: |
+    Thenables that reject with instances of the realm specified constructor function
+    satisfy the assertion, without cross realms collisions.
```
I was a bit confused by this description, it sounds like the opposite of what the test actually does.

cjtenny replied:
I copied it directly from assert-throws-same-realm.js. I *think* the description just needs to be updated in both, so I'll do that. But I'm not very familiar with realms yet.

ptomato commented on test/language/statements/class/decorator/syntax/valid/decorator-call-expr-identifier-reference-yield.js:
```diff
@@ -1,15 +1,19 @@
 // This file was procedurally generated from the following sources:
 // - src/decorator/decorator-call-expr-identifier-reference-yield.case
-// - src/decorator/syntax/valid/cls-decl-decorators-valid-syntax.template
+// - src/decorator/syntax/valid/cls-element-decorators-valid-syntax.template
```
This doesn't seem right. I remember somewhat recently there was a case where two test generators raced to produce the same file, which I believe was fixed. Is it possible that this branch is old enough that it doesn't have that fix, or maybe there's another similar case that wasn't caught yet?

cjtenny replied:
Yep, that was the case! A rebase, clean, and re-build fixed it. Thanks for catching it.

ptomato reviewed (changes requested):
Now that the RFC is merged I went over this in a bit more detail, and found a few things that should be fixed or clarified. Also it's slightly odd that the CI didn't run, so it would be good to find out why that is.

cjtenny commented:
Rebased and addressed feedback; feedback addressed in separate commit, nothing else changed in older commits. I'm sorry I dropped the ball on this and hadn't responded sooner.

ptomato commented on CONTRIBUTING.md:
```diff
@@ -344,52 +344,43 @@ Consumers that violate the spec by throwing exceptions for parsing errors at run
-}).then($DONE, $DONE);
+asyncTest(async function() {
+  assert.sameValue(true, await someTestCode(1), "someTestCode(1) should return true");
+;
```
```suggestion
```

ptomato commented on CONTRIBUTING.md:
```diff
@@ -344,52 +344,43 @@ Consumers that violate the spec by throwing exceptions for parsing errors at run
+
+ * If the argument to `$DONE` is a truthy value, the test is considered to have failed and the argument is displayed as the failure reason.
 
 ### Checking Exception Type and Message in Asynchronous Tests
```
```suggestion
### Checking Exception Type in Asynchronous Tests
```
We no longer have an example of how to check the exception message, but also I think we don't need one because I don't think any tests I've seen do so. Nor should they, since the contents of exception messages are not part of the requirements for a JS implementation.

ptomato reviewed (approved):
Just some minor loose ends!

Before merging would you mind debugging why the CI is not running? It looks like you may have to authorize the CircleCI app on your fork or something.
